File: CS12B/lab7/Dictionary.java
//------------------------------------------------------------------------------
// Sudeep Baniya 
// CMPS 12M
// sucbaniy
// 6/3/2016
// Binary Search Tree implementation of the Dictionary ADT
// Dictionary.java
//------------------------------------------------------------------------------

public class Dictionary implements DictionaryInterface {

   // private inner Node class
   private class Node {
      
      Pair item;
      
      Node left;
      
      Node right;

      Node(Pair x){
      
         item = x;
      
         left = null;
      
         right = null;
      }
   }

    // private inner Pair class
    private class Pair{
      
      String key;
      
      String value;
      
      Pair(String x, String y){
      
        key = x;
      
        value = y;
      }
    }

   // Fields for the DictionaryBST class
   private Node root;     // root = first node on list
   
   private int numItems;  // total number of items in the list

   // Dictionary()
   // constructor for the DictionaryBST class
   public Dictionary(){
   
      root = null;
   
      numItems = 0;
   }


   //helper functions ------------------------------------------------

   // findKey()
   // returns the Node containing key k in the subtree root, or returns NULL if no such Node exists
   private Node findKey(Node R, String targetKey){ 
   
      if(R == null || R.item.key.equals(targetKey)){
   
      return R;
      }
   
      if(R.item.key.compareToIgnoreCase(targetKey)>0){
   
      return findKey(R.left, targetKey);
   
      }else{
   
      return findKey(R.right, targetKey); 
   }
}
    
   // findParent()
   // returns the parent of N in the subtree rooted at R, or returns NULL if N is equal to R. (pre: R != NULL)
   Node findParent(Node N, Node R){
   
   Node P = null;
   
   if( N!=R ){
   
     P = R;
   
     while( P.left != N && P.right !=N ){
   
       if(N.item.key.compareToIgnoreCase(P.item.key)<0){
   
         P = P.left;
   
       }else{
   
         P = P.right;
     }
   }
   }
   
   return P;
   }

   // findLeftmost()
   // returns the leftmost Node in the subtree rooted at R, or NULL if R is NULL.
   Node findLeftmost(Node R){
   
   Node L = R;
   
   if( L!= null ) for( ; L.left != null; L = L.left);
   
   return L;
   
   }
   
   
   // printInOrder()
   // prints the (key, value) pairs belonging to the subtree rooted at R in order
   // of increasing keys to file pointed to by out.
   void printInOrder(Node R){
   
   if( R != null ){
   
      printInOrder(R.left);
   
      System.out.println(R.item.key + " " + R.item.value);
   
      printInOrder(R.right);
   }
   }
    
    // deleteAll()
    // deletes all Nodes in the subtree rooted at N.
    void deleteAll(Node N){
   
      if( N != null ){
   
      deleteAll(N.left);
   
      deleteAll(N.right);
      }
    }
    
    
   // BST ADT operations ----------------------------------------------------------

   // isEmpty()
   // pre: none
   // post: returns true if this Dictionary is empty, false otherwise
   public boolean isEmpty(){
   
      return(numItems == 0);
   
   }

   // size()
   // pre: none
   // returns the number of elements in this Dictionary
   public int size() {
   
      return numItems;
   }

   // lookup
   // pre: none
   // returns value associated key, or null reference if no such key exists
    public String lookup(String key){
   
      Node N;
   
      N = findKey(root, key);
   
      return ( N == null ? null : N.item.value );
    }


   // insert()
   // inserts new (key,value) pair into this Dictionary
   // pre: lookup(key)==null
    public void insert(String key, String value) throws DuplicateKeyException{
      
      Node N, A, B;
      
      if(findKey(root, key)!=null){
      
        throw new DuplicateKeyException("Dictionary Error: insert() cannot insert when duplicate keys occur");
      
      }
      
      N = new Node(new Pair(key, value));
      
      B = null;
      
      A = root;
      
      while( A != null ){
      
        B = A;
      
        if(A.item.key.compareToIgnoreCase(key)>0)  A = A.left;
      
       else A = A.right;
      
      }

      if( B == null ) root = N;
      
      else if(B.item.key.compareToIgnoreCase(key)>0)  B.left = N;
      
      else  B.right = N;
      
       numItems++;
    }
    

   // delete()
   // deletes pair with the given key
   // pre: lookup(key)!=null
    public void delete(String key) throws KeyNotFoundException{
      
      Node N, P, S;
      
      if(findKey(root, key)==null){
      
        throw new KeyNotFoundException("Dictionary Error: delete() cannot delete a key that does not exist");
      }
      
      N = findKey(root, key);
      
      if( N.left == null && N.right == null ){ // case 1
      
      if( N == root ){
      
        root = null;
      
      }else{
      
        P = findParent(N, root);
      
        if( P.right == N ) P.right = null;
      
        else P.left = null;
      
      }
      
      }else if( N.right == null ){ // case 2 (left but no right child) 
      
      if( N == root ){
      
        root = N.left;
      
      }else{
      
        P = findParent(N, root);
      
        if( P.right == N ) P.right = N.left;
      
        else P.left = N.left;
      
      }
      
      }else if( N.left == null ){   // case 2 (right but no left child)
      
      if( N == root ){
      
         root = N.right;
      
      }else{
      
         P = findParent(N, root);
      
         if( P.right == N ) P.right = N.right;
      
         else P.left = N.right;
      
      }
      
      }else{  // case 3: N.left!=null && N.right!=null
      
        S = findLeftmost(N.right);
      
        N.item.key = S.item.key;
      
        N.item.value = S.item.value;
      
        P = findParent(S, N);
      
        if( P.right == S ) P.right = S.right;
      
        else P.left = S.right; 
      
      }
      
      numItems--;
    }
        
   // makeEmpty
   // pre: none
    public void makeEmpty(){
      
      deleteAll(root);
      
      root = null;
      
      numItems = 0;
    }

   // toString()
   // Overrides Object's toString() method
   // pre: none
    public String toString(){
    
      printInOrder(root);
    
      return "";
    
    }
}

File: CS12B/pa5/Dictionary.c
//------------------------------------------------------------------------------
// Sudeep Baniya
// CMPS 12B
// sucbaniy
// 5-29-2016
// The programs implements a Dictionary ADT based on the hash table.
// Dictionary.c
//-------------------------------------------------------------------------------

#include<stdio.h>

#include<stdlib.h>

#include<string.h>

#include<assert.h>

#include"Dictionary.h"

#define MAX_LEN 180

const int tableSize = 101;

// hash functions ------------------------------------------------------------

// rotate_left()
// rotate the bits in an unsigned int
unsigned int rotate_left(unsigned int value, int shift) {
  
  int sizeInBits = 8*sizeof(unsigned int);

  shift = shift & (sizeInBits - 1);
  
  if ( shift == 0 )
  
    return value;
  
  return (value << shift) | (value >> (sizeInBits - shift));
}

// pre_hash()
// turn a string into an unsigned int
unsigned int pre_hash(char* input) { 
  
  unsigned int result = 0xBAE86554;
  
  while (*input) { 
  
    result ^= *input++;
  
    result = rotate_left(result, 5);
  }
  
  return result;
}

// hash()
// turns a string into an int in the range 0 to tableSize-1
int hash(char* key){
  
  return pre_hash(key)%tableSize;
}

// showBits()
// print out the bits in an unsigned int
void showBits(unsigned int x) {
  
  int i;
  
  int sizeInBits = 8*sizeof(unsigned int);
  
  char symbol[2] = {'0','1'};
  
  char* binary = malloc(sizeInBits + 1);

  memset(binary, 0, sizeInBits + 1);

  for (i=0; i<sizeInBits; i++) {
  
    binary[sizeInBits-i-1] = symbol[(x>>i) & 0x01];
  }
   
  printf("%s\n", binary);
  
  free(binary);
}

// private types --------------------------------------------------------------

// NodeObj
typedef struct NodeObj{
  
   char* key;
  
   char* value;
  
   struct NodeObj* next;

} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* k, char* v) {

   Node N = malloc(sizeof(NodeObj));

   assert(N!=NULL);

   N->key = k;

   N->value = v;

   N->next = NULL;

   return(N);
}

// freeNode()
// destructor for the Node type
void freeNode(Node* pN){

   if( pN!=NULL && *pN!=NULL ){

      free(*pN);

      *pN = NULL;
   }
}

// ListObj
typedef struct ListObj{

   Node first;

 } ListObj;

// List
typedef ListObj* List;

// newList()
// constructor of the Node type
List newList(void) {
 
   List L = malloc(sizeof(ListObj));
 
   assert(L!=NULL);
 
   L->first = NULL;
 
   return L;
}


// DictionaryObj
typedef struct DictionaryObj{

  List table;

  int numOfItems;

} DictionaryObj;


// private helper functions --------------------------------------------------

//deleteEverything
void deleteEverything(Node N){
  
  if (N!=NULL){
  
  deleteEverything(N->next);
  
    freeNode(&N);
  }
}

//findKey()
//returns a reference to the NodeObj containing its argument key or return null if no such NodeObj exists
Node findKey(Node A, char* targetKey){ 

while(A != NULL){

    if(strcmp(A->key, targetKey)==0){

      break;
    }

    A = A->next;
    }

    return A;
}


// public functions -----------------------------------------------------------

// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void){

   Dictionary D = malloc(sizeof(DictionaryObj));
  
   assert(D!=NULL);
  
   D->table = calloc(tableSize, sizeof(ListObj));
  
   D->numOfItems = 0;
  
   return D;
}

// Dictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD){
  
   if( pD!=NULL && *pD!=NULL ){
  
      if( !isEmpty(*pD) ) makeEmpty(*pD);

      free(*pD); 
	
      *pD = NULL;
   }
}

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D){
 
  if( D==NULL ){
 
      fprintf(stderr, "DICTIONARY Error: calling isEmpty() on NULL Dictionary reference\n");
 
      exit(EXIT_FAILURE);
   }
 
   return(D->numOfItems==0);
}

// size()
// returns the number of (key, value) pairs in D
// pre none
int size(Dictionary D){
 
  return D->numOfItems;
}

// lookup()
// returns the value v such that (k,v) is in D, or returns NULL if no 
// such value v exists
// pre: none
char* lookup(Dictionary D, char* k){
  
  int indexOftable;
  
  List L;
  
  Node N;
  
  indexOftable = hash(k);
  
  L = &D->table[indexOftable];
  
  N = findKey(L->first, k);
  
  if(N==NULL){
  
    return NULL;
  
  }else{
  
    return N->value;
    }
}

// insert()
// inserts new (key, value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char* k, char* v){
  
  List L;
  
  Node N;
  
  int indexOfTable;  
  
  indexOfTable = hash(k);
  
  L = &D->table[indexOfTable];
  
  if(findKey(L-> first, k)!=NULL){
  
    fprintf(stderr, "DICTIONARY Error: calling insert() when exists\n");
    
    exit(EXIT_FAILURE);
  }
  
  N = newNode(k, v);
  
  N->next = L->first;
  
  L->first = N;
  
  N = NULL;
  
  D->numOfItems++;
 
}


// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
void delete(Dictionary D, char* k){
  
  if( D==NULL ){
  
    fprintf(stderr, "DICTIONARY Error: calling delete() when nothing exists\n");
  
    exit(EXIT_FAILURE);
  }  
  
  List L;
  
  Node N;
  
  int indexOfTable;
  
  indexOfTable = hash(k);
  
  L = &D->table[indexOfTable];
  
  if(findKey(L->first, k)==L->first){
  
    N = L->first;
  
    L->first = L->first->next;
  
    N->next = NULL;
  
  }else{
  
    N = findKey(L->first, k);
  
    Node before = L->first;
  
    Node temporary = L->first->next;
    
    while(temporary !=N){
    
      temporary = temporary->next;
    
      before = before->next;
    }
    
    before->next = N->next;
    
    N->next = NULL;
  }
  
  D->numOfItems--;
  
  freeNode(&N);
}

// makeEmpty()
// re-sets D to the empty state
// pre: !isEmpty(D)
void makeEmpty(Dictionary D){
 
  List L;
 
  if( D==NULL ){
 
      fprintf(stderr, "DICTIONARY Error: calling makeEmpty() on NULL Dictionary reference\n");
 
      exit(EXIT_FAILURE);
   }
 
   if( D->numOfItems==0 ){
 
      fprintf(stderr, "DICTIONARY Error: calling makeEmpty() on empty Dictionary\n");
 
      exit(EXIT_FAILURE);
   }
   
   int i;
   
   for(i = 0; i<tableSize; i++){
 
       L = &D->table[i];
 
       deleteEverything(L->first);
   }
 
   D->numOfItems = 0; 

   D-> table = NULL;
   }

// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out
void printDictionary(FILE* out, Dictionary D){
   
   if( D==NULL ){
   
      fprintf(stderr, "DICTIONARY Error: calling printDictionary() on where the reference is NULL\n");
   
      exit(EXIT_FAILURE);
      }
  
  Node N;
  
  List L;
  
  int i;

  for(i = 0; i<tableSize; i++){
  
    L = &D->table[i];
  
    N = L->first;
  
    while(N!=NULL){
  
      fprintf(out, "%s %s\n", N->key, N->value);
  
      N = N->next;
    }
  }
}
